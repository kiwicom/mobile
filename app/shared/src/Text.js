// @flow

import * as React from 'react';
import ReactNative from 'react-native';

import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../types/Styles';

type Props = {|
  children: React.Node,
  style?: StylePropType,
|};

export default function Text({ children, style }: Props) {
  return (
    <ReactNative.Text style={[styles.nativeText, style]}>
      {children}
    </ReactNative.Text>
  );
}

const styles = StyleSheet.create({
  // These values are from the official design. Don't touch it please.
  nativeText: {
    fontWeight: 'normal',
    color: '#30363d',
    android: {
      fontSize: 15,
      lineHeight: 21,
    },
    ios: {
      fontSize: 14,
      lineHeight: 20,
    },
  },
});
