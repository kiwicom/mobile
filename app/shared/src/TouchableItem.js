// @flow

import * as React from 'react';
import {
  Platform,
  TouchableNativeFeedback,
  TouchableOpacity,
  View,
} from 'react-native';

import { type StylePropType } from '../types/Styles';

type Props = {
  // not exact: additional props allowed for TouchableOpacity and TouchableNativeFeedback
  children: React.Node,
  borderless?: boolean,
  pressColor?: string,
  style?: StylePropType,
};

const ANDROID_VERSION_LOLLIPOP = 21;

/**
 * TouchableItem renders a touchable that looks native on both iOS and Android.
 *
 * It provides an abstraction on top of TouchableNativeFeedback and
 * TouchableOpacity.
 *
 * On iOS you can pass the props of TouchableOpacity, on Android pass the props
 * of TouchableNativeFeedback.
 *
 * Note: this is basically copy-paste from React Navigation
 *
 * @see https://github.com/react-navigation/react-navigation/blob/cf36f22e68fe4a7f475d714590f665dfc29e3e25/src/views/TouchableItem.js
 */
export default class TouchableItem extends React.Component<Props> {
  static defaultProps = {
    borderless: false,
    pressColor: 'rgba(0, 0, 0, .32)',
  };

  render = () => {
    /*
     * TouchableNativeFeedback.Ripple causes a crash on old Android versions,
     * therefore only enable it on Android Lollipop and above.
     *
     * All touchables on Android should have the ripple effect according to
     * platform design guidelines.
     * We need to pass the background prop to specify a borderless ripple effect.
     */
    if (
      Platform.OS === 'android' &&
      Platform.Version >= ANDROID_VERSION_LOLLIPOP
    ) {
      const { style, ...rest } = this.props;
      return (
        <TouchableNativeFeedback
          useForeground={TouchableNativeFeedback.canUseNativeForeground()}
          {...rest}
          style={null}
          background={TouchableNativeFeedback.Ripple(
            this.props.pressColor,
            this.props.borderless,
          )}
        >
          <View style={style}>{React.Children.only(this.props.children)}</View>
        </TouchableNativeFeedback>
      );
    }

    // React.Children.only is only necessary for TouchableNativeFeedback
    // but since we are building for both platforms it is here as well
    // to discover this mistake as soon as possible
    return (
      <TouchableOpacity {...this.props}>
        {React.Children.only(this.props.children)}
      </TouchableOpacity>
    );
  };
}
