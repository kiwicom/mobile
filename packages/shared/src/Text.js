// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import ReactNative from 'react-native';
import { defaultTokens } from '@kiwicom/mobile-orbit';

import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../types/Styles';

// basically every component (but never strings - they must be translated)
// the idea is that you should always use component as a children so you'll
// eventually end up with the Translation component
type ChildrenType = React.Element<any>; // FIXME: we cannot accept View in the children because it's not supported on Android

type Props = {|
  +children: ChildrenType | React.ChildrenArray<ChildrenType>,
  +style?: StylePropType,
  +numberOfLines?: number,

  // none - View (Text in this case) is never the target of touch events
  +pointerEvents?: 'none',
|};

/**
 * React Native has the concept of style inheritance, but limited to text
 * subtrees. In this case, the second part will be both bold and red:
 *
 * ```
 * <Text style={{fontWeight: 'bold'}}>
 *   I am bold
 *   <Text style={{color: 'red'}}>and red</Text>
 * </Text>
 * ```
 *
 * It's roughly translated to this (`NSAttributedString` or `SpannableString`):
 *
 * ```
 * "I am bold and red"
 * 0-9: bold
 * 9-17: bold, red
 * ```
 *
 * This style inheritance works for deeply nested 'Text' components as well.
 *
 * Known issues:
 * @see https://github.com/facebook/react-native/issues/10712
 */
export default class Text extends React.Component<Props> {
  static childContextTypes = {
    isNested: PropTypes.bool,
  };

  static contextTypes = {
    isNested: PropTypes.bool,
  };

  getChildContext = () => ({
    isNested: true,
  });

  render() {
    let style = this.props.style;

    if (this.context.isNested === undefined) {
      // default styling must be applied only on the root Text component
      // `undefined` style in nested Text components indicates style inheritance
      style = [styleSheet.defaultTextStyle, this.props.style];
    }

    return (
      // $FlowFixMe Errors after moving rn modules from untyped to declarations
      <ReactNative.Text {...this.props} style={style}>
        {this.props.children}
      </ReactNative.Text>
    );
  }
}

const styleSheet = StyleSheet.create({
  // These values are from the official design. Don't touch it please.
  defaultTextStyle: {
    fontWeight: 'normal',
    color: defaultTokens.colorTextPrimary,
    android: {
      fontSize: 15,
      letterSpacing: 0,
      fontFamily: 'Roboto-Regular',
    },
    ios: {
      fontSize: 14,
      letterSpacing: -0.15,
      fontFamily: 'CircularPro-Book',
    },
  },
});
