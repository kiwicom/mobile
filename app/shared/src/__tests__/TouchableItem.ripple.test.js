// @flow

import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import Text from '../Text';
import TouchableItem from '../TouchableItem';

jest.mock('TouchableNativeFeedback', () => {
  const mock = jest.genMockFromModule('TouchableNativeFeedback');
  mock.canUseNativeForeground = () => true;
  mock.Ripple = jest.fn();
  return mock;
});
jest.mock('Platform', () => ({
  OS: 'android',
  Version: 22,
}));

const renderer = new ShallowRenderer();
const VoidAction = () => {};

describe('TouchableItem with ripple effect', () => {
  it('renders with foreground', () => {
    expect(
      renderer.render(
        <TouchableItem onPress={VoidAction}>
          <Text>line</Text>
        </TouchableItem>,
      ),
    ).toMatchSnapshot();
  });

  it('renders with ripple color', () => {
    expect(
      renderer.render(
        <TouchableItem rippleColor="red" onPress={VoidAction}>
          <Text>line</Text>
        </TouchableItem>,
      ),
    ).toMatchSnapshot();
  });

  it('disables foreground in borderless mode', () => {
    expect(
      renderer.render(
        <TouchableItem borderlessRipple={true} onPress={VoidAction}>
          <Text>line</Text>
        </TouchableItem>,
      ),
    ).toMatchSnapshot();
  });
});
