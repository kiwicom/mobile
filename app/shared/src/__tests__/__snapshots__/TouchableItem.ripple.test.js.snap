// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TouchableItem with ripple effect disables foreground in borderless mode 1`] = `
<DummyTouchableNativeFeedback
  background={undefined}
  borderlessRipple={true}
  onPress={[Function]}
  rippleColor="rgba(0, 0, 0, .32)"
  style={null}
  useForeground={false}
>
  <View
    style={undefined}
  >
    <Text>
      line
    </Text>
  </View>
</DummyTouchableNativeFeedback>
`;

exports[`TouchableItem with ripple effect renders with foreground 1`] = `
<DummyTouchableNativeFeedback
  background={undefined}
  borderlessRipple={false}
  onPress={[Function]}
  rippleColor="rgba(0, 0, 0, .32)"
  style={null}
  useForeground={true}
>
  <View
    style={undefined}
  >
    <Text>
      line
    </Text>
  </View>
</DummyTouchableNativeFeedback>
`;

exports[`TouchableItem with ripple effect renders with ripple color 1`] = `
<DummyTouchableNativeFeedback
  background={undefined}
  borderlessRipple={false}
  onPress={[Function]}
  rippleColor="red"
  style={null}
  useForeground={true}
>
  <View
    style={undefined}
  >
    <Text>
      line
    </Text>
  </View>
</DummyTouchableNativeFeedback>
`;
